http: add evhttp_connection_set_closecb to avoid g_requests hang

Crypt-iQ opened the pull request:
Prior to this patch, it was possible that if an RPC was interrupted before the reply was received, the success callback evhttp_request_set_on_complete_cb would never be called and g_requests wouldn't be cleaned up. When attempting to shutdown bitcoind, it would hang waiting for g_requests.empty().

Fixes https://github.com/bitcoin/bitcoin/issues/27722 

I am currently working on testing this with a libevent version of 2.2.1 to ensure that the patch also works for that version.

fanquake commented:
cc @fjahr

stickies-v reviewed (commented):
Concept ACK.

Was able to [reproduce the bug](https://github.com/bitcoin/bitcoin/issues/27722#issuecomment-1559453982), and this fixes the issue for me (libevent 2.1.12, macOS 13.4).

Need to read up on the libevent workaround a bit more first.

promag commented:
Why I am not co-author?

Crypt-iQ commented:
> Why I am not co-author?

Happy to add you of course since you wrote it -- I couldn't find your email via `git log` to add it and I wasn't sure how to give credit w/o an email

Crypt-iQ commented:
Can you advise on how to do that? I can close this in favor and let somebody take it over with proper commit attribution. I opened this PR since it wasn't addressed for a month

borgbyte commented:
yup that’s my email. just add a new co-authored line?

Crypt-iQ commented:
> git log for me has @promag's email as João Barbosa <joao.paulo.barbosa@gmail.com>

I see, I grepped for `promag` instead of the name. Anyways, the commit attribution is fixed. Sorry about that @promag

promag commented:
@Crypt-iQ No worries mate, thanks for picking this work!

stickies-v reviewed (commented):
I agree with the approach of calling `evhttp_connection_set_closecb` regardless of whether `EV_READ` is enabled or disabled, which is currently the only difference between this PR and https://github.com/bitcoin/bitcoin/pull/27245.

With this PR:
- In the case that `EV_READ` is enabled, `evhttp_connection_set_closecb` will be called as soon as the connection is closed, including in case of a remote client disconnection (especially useful during long-running requests as described [here](https://github.com/bitcoin/bitcoin/issues/27722#issuecomment-1559453982)), which is the most efficient as we're not wasting server resources on completing the request that won't get served anyway.

- In the case that `EV_READ` is disabled, `evhttp_connection_set_closecb` will not be called immediately in case of remote client disconnection (because we're not listening for it), but it will still be called when the server is shutting down, still resolving the issue in https://github.com/bitcoin/bitcoin/pull/27909 where bitcoind can't gracefully shutdown but potentially wasting resources on requests of which the connection has been closed.

However, I'm not sure the [entire code block](https://github.com/bitcoin/bitcoin/pull/27909/files#diff-63c8cb9c9dd61d50d59afd5c39914e1c259f8743030b637a7896a0746c851ef1R221-R229) to enable `EV_READ` is safe, as per my [comment](https://github.com/bitcoin/bitcoin/pull/27909#discussion_r1238859568). I'll continue digging into that but looking forward to hearing other views on this? If it's not safe, I think we need to leave this out and look for alternatives to monitoring remote client disconnections. A fix is suggested in https://github.com/azat/libevent/commit/7d2e2a84ce7bab938433851ff4e842e15c005dc4, but I [can't see it in the 2.2.1-alpha release yet](https://github.com/libevent/libevent/blob/release-2.2.1-alpha/http.c#L1147). Even with `EV_READ` disabled, I still think calling `evhttp_connection_set_closecb` is a very worthwhile improvement.

Crypt-iQ commented on src/httpserver.cpp:
```diff
@@ -215,6 +215,30 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
+        // The EV_READ enablement is off-set by the workaround for the libevent
+        // bug further below. This is why it can only be used with libevent
+        // versions that are not affected by the workaround code.
+        if (event_get_version_number() < 0x02010600 && event_get_version_number() >= 0x02020001) {
```
Yes that's right

Crypt-iQ commented on src/httpserver.cpp:
```diff
@@ -215,6 +215,30 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
+            // Enable EV_READ to detect remote disconnection meaning that the close
+            // callback set below is called.
+            bufferevent* bev = evhttp_connection_get_bufferevent(conn);
+            bufferevent_enable(bev, EV_READ);
```
I wasn't able to reproduce using this https://github.com/bitcoin/bitcoin/pull/11593#issuecomment-341489496 and since you pointed out that the change wasn't merged into 2.2, I think it's better to err on the side of caution and not enable reading

Crypt-iQ replied:
I've updated the PR and removed this

fjahr commented:
> I opened this PR since it wasn't addressed for a month

Why didn't you ask me first if I am still working on this? A month isn't a long time in this project when you working on multiple things at the same time...

fjahr commented:
@Crypt-iQ There is no need to close this PR now since it got attention and review. I will close mine if you want to continue. But I wanted to clarify this for the future.

Crypt-iQ commented:
Ok I reopened, it would be helpful if these things were made explicit (leave a comment) rather than implicit

stickies-v commented on src/httpserver.cpp:
```diff
@@ -215,6 +215,30 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
+        // The EV_READ enablement is off-set by the workaround for the libevent
+        // bug further below. This is why it can only be used with libevent
+        // versions that are not affected by the workaround code.
+        if (event_get_version_number() < 0x02010600 && event_get_version_number() >= 0x02020001) {
```
Opened a pull to correct this: https://github.com/bitcoin/bitcoin/pull/27949

promag committed:
http: add evhttp_connection_set_closecb to avoid g_requests hang

Prior to this patch, it was possible that if an RPC was interrupted
before the reply was received, the success callback
evhttp_request_set_on_complete_cb would never be called and
g_requests wouldn't be cleaned up. When attempting to shutdown
bitcoind, it would hang waiting for g_requests.empty().

Co-authored-by: Fabian Jahr <fjahr@protonmail.com>

Crypt-iQ commented:
> @Crypt-iQ There is no need to close this PR now since it got attention and review. I will close mine if you want to continue. But I wanted to clarify this for the future.

I think you can reopen your PR to include the bufferevent read enable. I've updated this PR to include a minimal fix for the issue (without the bufferevent read enable) because I'm still a little unclear on what exactly the bug fix was originally for and I can't reproduce it. Also happy to close this PR in favor of yours since this is now missing the read enable.

stickies-v reviewed (commented):
Approach ACK 710985d6953eb73b5edddaa7d4d54b58c90d1d34

I think improving remote client disconnect detection is best left for a separate PR; this is a worthwhile improvement on its own and I think uncontroversial.

Crypt-iQ commented on src/httpserver.cpp:
```diff
@@ -215,6 +215,20 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
+            auto req = static_cast<evhttp_request*>(arg);
+            LOCK(g_requests_mutex);
+            auto n{g_requests.erase(req)};
+            if (n == 1 && g_requests.empty()) g_requests_cv.notify_all();
```
I think because the only call to `g_requests_cv.wait(...)` has a predicate that waits for `g_requests.empty()`. I've removed the dependency in the latest patch

Crypt-iQ commented:
I've removed the `assert(n == 1)` in `evhttp_request_set_on_complete_cb` because of this comment: https://github.com/bitcoin/bitcoin/issues/27722#issuecomment-1611979215. It appears that both `evhttp_connection_set_closecb` and `evhttp_request_set_on_complete_cb` can run which causes a crash. I wasn't able to reproduce it and still want to know what exactly happened, but I think removing the assert here is certainly better than having a crash.

Crypt-iQ commented on src/httpserver.cpp:
```diff
@@ -215,6 +215,20 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
         WITH_LOCK(g_requests_mutex, g_requests.insert(req));
         g_requests_cv.notify_all();
+
+        auto conn = evhttp_request_get_connection(req);
```
done

Crypt-iQ commented on src/httpserver.cpp:
```diff
@@ -215,6 +215,20 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
+        // called if the connection is closed after a successful request, but
+        // complete callback set below already cleared the state.
+        evhttp_connection_set_closecb(conn, [](evhttp_connection* conn, void* arg) {
+            auto req = static_cast<evhttp_request*>(arg);
```
done

stickies-v commented:
> I wasn't able to reproduce it and still want to know what exactly happened, but I think removing the assert here is certainly better than having a crash.

I'm able to reproduce @dooglus's observed crash quite quickly (< 1 second usually) by blasting the server with requests from > 1 thread.

<details>
<summary>diff and script to make it crash</summary>

On top of d170e6c60cd9f46da1794448b8be55ba6f0b2922, first re-add the assertion:

```diff
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index 849e9b482..5683e761f 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -229,7 +229,8 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
 
         // Clear state after successful request.
         evhttp_request_set_on_complete_cb(req, [](struct evhttp_request* req, void*) {
-            WITH_LOCK(g_requests_mutex, g_requests.erase(req));
+            auto n{WITH_LOCK(g_requests_mutex, return g_requests.erase(req))};
+            assert(n == 1);
             g_requests_cv.notify_all();
         }, nullptr);
     }

```

I'm running on signet with `user:pass` credentials:

```sh
./src/bitcoind -signet -debug=http -rpcuser=user -rpcpassword=pass
```

And then finally the script to make it crash:
```py
import http.client
import base64
from concurrent.futures import ThreadPoolExecutor

host = "localhost"
port = 38332

auth_string = base64.b64encode('user:pass'.encode()).decode()

num_threads = 2
iterations_per_thread = 10000

def send_requests(i):
    for _ in range(iterations_per_thread):
        conn = http.client.HTTPConnection(host, port)
        conn.request("POST", "/")
        response = conn.getresponse()

        conn.close()

with ThreadPoolExecutor(max_workers=num_threads) as executor:
    executor.map(send_requests, range(num_threads))
```
</details>

I agree that keeping the assertion in as is is bad, but I don't like just removing it without properly understanding why this fails (or probably in other words, why/when the `evhttp_connection_set_closecb` cb is called before the `evhttp_request_set_on_complete_cb` cb) either, so I think we need to investigate first before proceeding?

Crypt-iQ commented:
I am a complete libevent noob, but I noticed some odd behavior that makes me think this is a race condition. I used the following diff that logged both sides' port numbers and the above python script:

<details>
<summary>diff</summary>
<br>

```
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index 849e9b482..8bb9d407f 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -223,13 +223,21 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
         // complete callback set below already cleared the state.
         evhttp_connection_set_closecb(conn, [](evhttp_connection* conn, void* arg) {
             auto req{static_cast<evhttp_request*>(arg)};
+	    auto sockd = (struct sockaddr_in*)evhttp_connection_get_addr(conn);
+	    auto addr = sockd->sin_port;
+	    LogPrint(BCLog::HTTP, "close callback remote port %d conn port %d\n", req->remote_port, addr);
             WITH_LOCK(g_requests_mutex, g_requests.erase(req));
             g_requests_cv.notify_all();
         }, req);
 
         // Clear state after successful request.
         evhttp_request_set_on_complete_cb(req, [](struct evhttp_request* req, void*) {
-            WITH_LOCK(g_requests_mutex, g_requests.erase(req));
+	    auto conn = req->evcon;
+	    auto sockd = (struct sockaddr_in*)evhttp_connection_get_addr(conn);
+	    auto addr = sockd->sin_port;
+            LogPrint(BCLog::HTTP, "complete callback remote port %d conn port %d\n", req->remote_port, addr);
+	    auto n{WITH_LOCK(g_requests_mutex, return g_requests.erase(req))};
+	    assert(n == 1);
             g_requests_cv.notify_all();
         }, nullptr);
     }
```

</details>

Most of the time (98-99%), I would see normal log lines where both ports were the same in the complete and close callbacks:

```
2023-07-16T22:25:46Z [http] Received a POST request for / from 127.0.0.1:44596
2023-07-16T22:25:46Z [http] complete callback remote port 44596 conn port 13486
2023-07-16T22:25:46Z [http] close callback remote port 44596 conn port 13486
```

However, every time the assert failure was hit, the port numbers were not the same for the complete/close callbacks. The port numbers were also re-used from another ongoing request. In the below snippet, the close callback for the request using remote port 44688 uses the local port 36526 which also matches the local port used in the complete callback for remote port 44686. Also, the complete callback for remote port 44688 has a local port 37038 instead of 36526.

```
2023-07-16T22:25:46Z [http] close callback remote port 44686 conn port 34990
2023-07-16T22:25:46Z [http] Received a POST request for / from 127.0.0.1:44686
2023-07-16T22:25:46Z [http] complete callback remote port 44686 conn port 36526
2023-07-16T22:25:46Z [http] Received a POST request for / from 127.0.0.1:44688
2023-07-16T22:25:46Z [http] close callback remote port 44688 conn port 36526
2023-07-16T22:25:46Z [http] complete callback remote port 44688 conn port 37038
<crash here>
```

Crypt-iQ commented:
I modified the diff above and printed out the `evhttp_request` memory address as well as the `output_buffer` memory address. I noticed that the`evhttp_request` address is re-used when there's a crash and the `output_buffer` address is not consistent between callbacks.

<details>
<summary>diff</summary>
<br>

```
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index 849e9b482..9fefad6bd 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -211,6 +211,7 @@ std::string RequestMethodString(HTTPRequest::RequestMethod m)
 /** HTTP request callback */
 static void http_request_cb(struct evhttp_request* req, void* arg)
 {
+    LogPrint(BCLog::HTTP, "Received a request from %d req %p\n", req->remote_port, (void*)req);
     // Track requests and notify when a request is completed.
     {
         WITH_LOCK(g_requests_mutex, g_requests.insert(req));
@@ -222,14 +223,22 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
         // called if the connection is closed after a successful request, but
         // complete callback set below already cleared the state.
         evhttp_connection_set_closecb(conn, [](evhttp_connection* conn, void* arg) {
-            auto req{static_cast<evhttp_request*>(arg)};
+	     auto req{static_cast<evhttp_request*>(arg)};
+            auto sockd = (struct sockaddr_in*)evhttp_connection_get_addr(conn);
+	     auto addr = sockd->sin_port;
+	     LogPrint(BCLog::HTTP, "set_closecb called for remote port %d conn port %d req %p output buffer %p\n", req->remote_port, addr, (void*)req, (void*)req->output_buffer);
             WITH_LOCK(g_requests_mutex, g_requests.erase(req));
             g_requests_cv.notify_all();
         }, req);
 
         // Clear state after successful request.
         evhttp_request_set_on_complete_cb(req, [](struct evhttp_request* req, void*) {
-            WITH_LOCK(g_requests_mutex, g_requests.erase(req));
+	     auto conn = req->evcon;
+	     auto sockd = (struct sockaddr_in*)evhttp_connection_get_addr(conn);
+	     auto addr = sockd->sin_port;
+            LogPrint(BCLog::HTTP, "set_on_complete_cb called for remote port %d conn port %d req %p output buffer %p\n", req->remote_port, addr, (void*)req, (void*)req->output_buffer);
+            auto n{WITH_LOCK(g_requests_mutex, return g_requests.erase(req))};
+            assert(n == 1);
             g_requests_cv.notify_all();
         }, nullptr);
     }

```
</details>

In the following log snippet, the request from remote ports 56815 and 56816 use the same memory address `0x600000224000`. When the close callback is called, I would have expected an `output_buffer` address of `0x6000009301b0` rather than `0x6000009302d0`. Note that `g_requests` is a set of `evhttp_request*`.  So the first complete callback removes the request with address `0x600000224000`. Then a new request is received with the same address and is inserted into `g_requests`. Then the close callback runs (I'm not sure which request this close callback belongs to) and removes the entry with address `0x600000224000`. Then the last complete callback runs and the entry has already been deleted.

```
2023-07-17T11:50:09Z [http] Received a request from 56815 req 0x600000224000
2023-07-17T11:50:09Z [http] Received a POST request for / from [::1]:56815
2023-07-17T11:50:09Z [http] set_on_complete_cb called for remote port 56815 conn port 61405 req 0x600000224000 output buffer 0x6000009301b0
2023-07-17T11:50:09Z [http] Received a request from 56816 req 0x600000224000
2023-07-17T11:50:09Z [http] Received a POST request for / from [::1]:56816
2023-07-17T11:50:09Z [http] set_closecb called for remote port 56816 conn port 61405 req 0x600000224000 output buffer 0x6000009302d0
2023-07-17T11:50:09Z [http] set_on_complete_cb called for remote port 56816 conn port 61661 req 0x600000224000 output buffer 0x6000009302d0
Assertion failed: (n == 1), function operator(), file httpserver.cpp, line 241.
Abort trap: 6
```

stickies-v commented on src/httpserver.cpp:
```diff
@@ -215,9 +215,21 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
+            auto req{static_cast<evhttp_request*>(arg)};
+            WITH_LOCK(g_requests_mutex, g_requests.erase(req));
+            g_requests_cv.notify_all();
+        }, req);
```
This is the culprit of [what you're observing](https://github.com/bitcoin/bitcoin/pull/27909#issuecomment-1637992727). `req` becomes a dangling pointer as soon as a request is completed. In most cases (which is why we typically need > 1 number of iterations), this just means that `g_requests.erase(req)` removes 0 elements because it points to a non-existing request, but sometimes (and actually quite frequently as addresses seem to be reused quite aggressively) this will point to an unrelated request which has already been added into `g_requests`, erasing it unintentionally.

The underlying issue is that an `evhttp_connection` and an `evhttp_request` simply have different lifecycles, with connections living longer than requests. It seems that passing `req` as an argument to `evhttp_connection_set_closecb` is a fundamentally unsafe thing to do if we can't guarantee that `req` hasn't been completed by the time `closecb` is called (which in the happy path, we always expect it to be).

Crypt-iQ replied:
nice catch!

vasild replied:
What about this:

```cpp
evhttp_connection_set_closecb(
    evhttp_request_get_connection(req),
    [](evhttp_connection* conn, void* arg) {
        auto req{static_cast<evhttp_request*>(arg)};
        // By the time the connection close callback is executed (this lambda)
        // the request we scheduled to be passed as an argument (arg) may
        // have been destroyed and a new, unrelated, one could have been
        // created at the same address. So arg may be a stale pointer or
        // a valid pointer, pointing to a new request object.
        LOCK(g_requests_mutex);
        if (g_requests.count(req) > 0 /* is not a stale pointer */
            && evhttp_request_get_connection(req) == conn /* is not pointing to a new request */) {

            g_requests.erase(req);
            g_requests_cv.notify_all();
        }
    },
    req);
```

stickies-v replied:
```
            && evhttp_request_get_connection(req) == conn /* is not pointing to a new request */) {
```

I'm not sure this will be 100% robust when we have multiple requests in a connection, I think this can still lead to usage of stale pointers? But as we're closing the connection anyway, it may not be an issue.

stickies-v replied:
I think the most sensible approach is to instead of track requests, keep a per-connection request counter. It's a pretty straightforward approach with a `std::unordered_map<const evhttp_connection*, size_t>`, but I've added a `RequestTracker` helper class wrapping around that to make that a bit more robust. What do you think?

https://github.com/stickies-v/bitcoin/commit/953c6914806bdaa6d087984049dccfe180dffc20

stickies-v replied:
I've opened https://github.com/bitcoin/bitcoin/pull/28551 with this approach to try and get it in before v26.

stickies-v reviewed (commented):
Nice investigative work, think I found the issue as per [my comment](https://github.com/bitcoin/bitcoin/pull/27909#discussion_r1265614783). 

I wonder if we could track connections instead of requests. A quick implementation (docstrings etc generally not updated, just a PoW) seems to indicate that it does:

<details>
<summary>git diff on d170e6c60cd9f46da1794448b8be55ba6f0b2922</summary>

```diff
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index 849e9b482..8fe826f55 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -151,9 +151,9 @@ static std::vector<HTTPPathHandler> pathHandlers GUARDED_BY(g_httppathhandlers_m
 //! Bound listening sockets
 static std::vector<evhttp_bound_socket *> boundSockets;
 //! Track active requests
-static GlobalMutex g_requests_mutex;
-static std::condition_variable g_requests_cv;
-static std::unordered_set<evhttp_request*> g_requests GUARDED_BY(g_requests_mutex);
+static GlobalMutex g_http_conns_mutex;
+static std::condition_variable g_http_conns_cv;
+static std::unordered_set<evhttp_connection*> g_http_conns GUARDED_BY(g_http_conns_mutex);
 
 /** Check if a network address is allowed to access the HTTP server */
 static bool ClientAllowed(const CNetAddr& netaddr)
@@ -213,25 +213,16 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
 {
     // Track requests and notify when a request is completed.
     {
-        WITH_LOCK(g_requests_mutex, g_requests.insert(req));
-        g_requests_cv.notify_all();
-
         auto conn{evhttp_request_get_connection(req)};
+        WITH_LOCK(g_http_conns_mutex, g_http_conns.insert(conn));
+        g_http_conns_cv.notify_all();
 
-        // Close callback to clear active but running request. This is also
-        // called if the connection is closed after a successful request, but
-        // complete callback set below already cleared the state.
-        evhttp_connection_set_closecb(conn, [](evhttp_connection* conn, void* arg) {
-            auto req{static_cast<evhttp_request*>(arg)};
-            WITH_LOCK(g_requests_mutex, g_requests.erase(req));
-            g_requests_cv.notify_all();
-        }, req);
 
-        // Clear state after successful request.
-        evhttp_request_set_on_complete_cb(req, [](struct evhttp_request* req, void*) {
-            WITH_LOCK(g_requests_mutex, g_requests.erase(req));
-            g_requests_cv.notify_all();
-        }, nullptr);
+        // Remove connection when it is closed
+        evhttp_connection_set_closecb(conn, [](evhttp_connection* conn, void* arg) {
+            auto n{WITH_LOCK(g_http_conns_mutex, return g_http_conns.erase(conn))};
+            g_http_conns_cv.notify_all();
+        }, NULL);
     }
 
     // Disable reading to work around a libevent bug, fixed in 2.2.0.
@@ -486,12 +477,12 @@ void StopHTTPServer()
     }
     boundSockets.clear();
     {
-        WAIT_LOCK(g_requests_mutex, lock);
-        if (!g_requests.empty()) {
-            LogPrint(BCLog::HTTP, "Waiting for %d requests to stop HTTP server\n", g_requests.size());
+        WAIT_LOCK(g_http_conns_mutex, lock);
+        if (!g_http_conns.empty()) {
+            LogPrint(BCLog::HTTP, "Waiting for %d connections to stop HTTP server\n", g_http_conns.size());
         }
-        g_requests_cv.wait(lock, []() EXCLUSIVE_LOCKS_REQUIRED(g_requests_mutex) {
-            return g_requests.empty();
+        g_http_conns_cv.wait(lock, []() EXCLUSIVE_LOCKS_REQUIRED(g_http_conns_mutex) {
+            return g_http_conns.empty();
         });
     }
     if (eventHTTP) {

```
</details>

This assumes that a connection can never have more than 1 request, and I'm not sure if that's the case.

Crypt-iQ commented:
> This assumes that a connection can never have more than 1 request, and I'm not sure if that's the case.

I'm also not sure. This PR body leads me to believe that it is possible to have more than 1 request per connection: https://github.com/libevent/libevent/pull/592#issue-292981989 , however I'm not sure whether it's possible for them to be interleaved and cause problems with your connection-based patch. Something I'll try to investigate.

Crypt-iQ commented:
It is possible for a connection to have more than one request (see: https://github.com/libevent/libevent/issues/1383#issuecomment-1383249281) or `evhttp_associate_new_request_with_connection` in the libevent code. It appears though that the requests are sequential so when one completes the next one can run. I modified the python script to test this, but it doesn't lead to a deadlock or crash because when the next request comes in, it uses the same `evhttp_connection*` and so the insert into `g_http_conns` is a no-op. The next request will also update the callback on the `evhttp_connection*` via `evhttp_connection_set_closecb` so that only one callback is run. When the connection is finally closed, it will find the connection & delete it. I think this means your patch works, but I'll continue to investigate a little more

<details>
<summary>python script</summary>
<br>

```python
import http.client
import base64
from concurrent.futures import ThreadPoolExecutor
import time

host = "localhost"
port = 18443

auth_string = base64.b64encode('user:pass'.encode()).decode()

num_threads =  1
iterations_per_thread = 1

def send_requests(i):
    for _ in range(iterations_per_thread):
        conn = http.client.HTTPConnection(host, port)
        conn.request("POST", "/", headers={"Keep-Alive": "timeout=500, max=1000"})
        
        #response = conn.getresponse()
        #conn.close()

        time.sleep(3)

        conn.request("POST", "/", headers={"Keep-Alive": "timeout=500, max=1000"})

        time.sleep(100)

with ThreadPoolExecutor(max_workers=num_threads) as executor:
    executor.map(send_requests, range(num_threads))
```
</details>

Crypt-iQ commented:
I haven't been able to cause a crash with the connection-based code. Given that the connection-based code is different than the crashing code here, do you want to open a new PR with the connection-based code @stickies-v ?

fanquake commented:
What is that status of this? cc @stickies-v.

vasild commented:
https://github.com/libevent/libevent/issues/78 is relevant. Need to check how libevent was fixed to detect client close and make use of that in `bitcoind`.

stickies-v commented:
> Need to check how libevent was fixed to detect client close and make use of that in `bitcoind`.

As far as I could find, it's not yet fixed but it's also largely orthogonal, see my comment [here](https://github.com/bitcoin/bitcoin/pull/27909#pullrequestreview-1493733432) and [this open libevent issue](https://github.com/libevent/libevent/issues/666). Detecting a remote disconnect would help save resources by not completing a request we'll no longer be serving, but isn't necessary to avoid the hangs on shutdown.

fanquake commented:
Going to mark this as draft for now, given that #28551 is open.

stickies-v commented:
Superseded by #28551, can be closed now.

fanquake commented:
Closing for now.
